In [47]:
# Installing the Pyspark to handle Big Data
!pip install pyspark -q

In [48]:
# Connect Google drive with collab
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [49]:
from pyspark.sql import SparkSession

spark = SparkSession.builder\
        .appName("NYC_Taxi_BigData")\
        .config('spark.sql.shuffle.partitions', '200')\
        .getOrCreate()

print("Spark Started Successfully")


Spark Started Successfully


In [50]:
# We will check the version of Spark
spark.version

'4.0.2'

In [51]:
import os
os.listdir('/content/drive/MyDrive')

['Colab Notebooks',
 'IMG-20221209-WA0017 (1).jpg',
 'IMG-20221209-WA0016.jpg',
 'QFIX-PAYMENT-RECEIPT-Hostel Fee Odd Semester Sixty Percentage.pdf',
 '22EDACS071.pdf',
 '20230907_215427.jpg',
 '20210119_160705.jpg',
 'IMG-20221209-WA0017.jpg',
 '20231107_125357 (1).jpg',
 '20210116_102904.jpg',
 '20231107_125357.jpg',
 '20231107_125408.jpg',
 '20221012_151945 (1).jpg',
 '20230909_120245 (1).jpg',
 '20230909_120245.jpg',
 '20221012_151945.jpg',
 'Sahil Sharma.png',
 '20210131_174613.jpg',
 'IMG_20240112_184350421~2.jpg',
 'Screenshot_20240112-220414.png',
 'Screenshot_20240112-221357~2 (1).png',
 'Screenshot_20240112-221357~2.png',
 'Mathematics unit-1.pdf',
 '2019.pdf',
 '2023.pdf',
 '2022.pdf',
 'Adobe Scan 26-Dec-2023.pdf',
 'DE notes.pdf',
 'SE unit 5-1.pdf',
 'SE unit 4-1.pdf',
 'SE unit 3-1.pdf',
 'SE unit 2.pdf',
 'SE Unit 1.pdf',
 'Developing a Remote.pdf',
 '20230907_234409.mp4',
 'download.jpeg',
 'PPT Presentation SIH 2024.gdoc',
 'Final PPT.pptx',
 'SiH_Project_2024[1].pptx

In [52]:
os.listdir('/content/drive/MyDrive/Ml Assignment CBI Solutions')

['data', 'outputs', 'models']

In [53]:
os.listdir('/content/drive/MyDrive/Ml Assignment CBI Solutions/data')[:5]

['yellow_tripdata_2024-03.parquet',
 'yellow_tripdata_2024-04.parquet',
 'yellow_tripdata_2024-05.parquet',
 'yellow_tripdata_2024-06.parquet',
 'yellow_tripdata_2024-07.parquet']

In [54]:
df = spark.read.parquet(
    "/content/drive/MyDrive/Ml Assignment CBI Solutions/data/yellow_tripdata_2024-*.parquet"
)

In [55]:
print("Number of Columns:", len(df.columns))

Number of Columns: 19


In [56]:
print("Total Rows:",df.count())

Total Rows: 41169720


In [57]:
df.show(5)

+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+
|       2| 2024-10-01 00:30:44|  2024-10-01 00:48:26|              1|          3.0|         1|                 N|         162|         246|           1|       18.4|  1.0|    0.5|       1.

In [58]:
df.printSchema() # Gives me column with thier data types.

root
 |-- VendorID: integer (nullable = true)
 |-- tpep_pickup_datetime: timestamp_ntz (nullable = true)
 |-- tpep_dropoff_datetime: timestamp_ntz (nullable = true)
 |-- passenger_count: long (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: long (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- payment_type: long (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- Airport_fee: double (nullable = true)



# **Summary about my columns :**

1. VendorID – Taxi company ID.
2. tpep_pickup_datetime – Trip start date and time.
3. tpep_dropoff_datetime – Trip end date and time.
4. passenger_count – Number of passengers.
5. trip_distance – Distance traveled during the trip.
6. RatecodeID – Fare rate type.
7. store_and_fwd_flag – Data stored before transmission (Y/N).
8. PULocationID – Pickup location ID.
9. DOLocationID – Drop-off location ID.
10. payment_type – Payment method used.
11. fare_amount – Base fare of the trip.
12. extra – Additional charges.
13. mta_tax – MTA tax amount.
14. tip_amount – Tip paid by the passenger.
15. tolls_amount – Toll charges.
16. improvement_surcharge – Additional surcharge.
17. total_amount – Total trip amount.
18. congestion_surcharge – Congestion charge.
19. Airport_fee – Airport-related fee.

# Dataset Information

1. Dataset Name: NYC Yellow Taxi Trip Records
2. Total Rows: 41,169,720
3. Total Columns: 19
4. File Format: Parquet
5. Processing Tool: Apache Spark 4.0.2

In [59]:
df.show(5, truncate=False)

+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+
|2       |2024-10-01 00:30:44 |2024-10-01 00:48:26  |1              |3.0          |1         |N                 |162         |246         |1           |18.4       |1.0  |0.5    |1.5      

In [60]:
# Checking missing values:
from pyspark.sql.functions import col, sum
missing_df = df.select([
    sum(col(c).isNull().cast("int")).alias(c)
    for c in df.columns
])

missing_df.show()

+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+
|       0|                   0|                    0|        4091232|            0|   4091232|           4091232|           0|           0|           0|          0|    0|      0|         

# **Summary for Missing values**

1. Missing values were found in the following columns:
   - passenger_count
   - RatecodeID
   - store_and_fwd_flag
   - congestion_surcharge
   - Airport_fee

2. All remaining columns had no missing values.
3. Handel misiing values before model training

**Extra Points**

Dataset mein total rows - 41,169,720 rows
and in rows - 4,091,232(4.09 million records) , there are some columns which are missing

Percentage of missing data = (4,091,232 / 41,169,720) × 100 ≈ 9.94%

10% data missing hai total data mein se

# **Data Cleaning**

In [61]:
# STatiscal Summarry dega for numerical COlumns.
df.select(
    'trip_distance',
    'fare_amount',
    'passenger_count'
).describe().show()

+-------+-----------------+------------------+------------------+
|summary|    trip_distance|       fare_amount|   passenger_count|
+-------+-----------------+------------------+------------------+
|  count|         41169720|          41169720|          37078488|
|   mean| 4.97610078523786|19.268509729958033|1.3339306338489314|
| stddev|419.2304973495382| 76.71984026978836|0.8158242156146562|
|    min|              0.0|           -2261.2|                 0|
|    max|        398608.62|         335544.44|                 9|
+-------+-----------------+------------------+------------------+



# Initial Data Observations

1. The average trip distance is 4.98 miles.

2. The average fare amount is $19.27.

3. The average passenger count is 1.33.

4. Some records contain unrealistic values such as:
   - Negative fare amounts : -2261.2
   - Zero passenger count : 0
   - Extremely large trip distances :  398608.62 (😂)
   - Extremely large fare amounts :  335544.44 (😅)

5. Yeh sab outliers and will be removed during the data cleaning phase.

In [62]:
df.filter(df.fare_amount <= 0).count()

748284

# Fare Amount Outlier Check

fare_amount <= 0 records = 748,284

Total rows = 41,169,720

Fare = 0,-ve = (748,284 / 41,169,7200) × 100 ≈ 1.82%

Since -ve data zyada nhi hai to we will remove these rows


In [63]:
df.filter(df.trip_distance <= 0).count()

776305

# Trip Distance Outlier Check

trip_distance <= 0:  records = 776305

Total rows = 41,169,720

trip_distance = 0,-ve = (776305 / 41,169,7200) × 100 ≈ 1.88%

Since -ve data zyada nhi hai to we will remove these rows


In [64]:
df.filter(df.passenger_count <= 0).count()

401354

# Passenger Count Outlier Check

passenger_count <= 0:  records = 401354

Total rows = 41,169,720

passenger_count = 0,-ve = (401354 / 41,169,7200) × 100 ≈ 0.97%

Since -ve data zyada nhi hai to we will remove these rows


In [65]:
# removing the inavlid rows :
clean_df = df.filter(
    (df.fare_amount > 0) &
    (df.trip_distance > 0) &
    (df.passenger_count > 0)
)

In [66]:
print("Original Rows:", df.count())
print("Cleaned Rows:", clean_df.count())
print('Removed Rows:', df.count() - clean_df.count())

Original Rows: 41169720
Cleaned Rows: 35629110
Removed Rows: 5540610


# Data Cleaning Summary

1. Invalid records were removed from the dataset.

2. The following cleaning rules were applied:
   - fare_amount > 0
   - trip_distance > 0
   - passenger_count > 0

3. Original number of records: 4,11,69,720 (4.1 cr)

4. Number of records after cleaning: 3,56,29,110 (3.5 cr)

5. Total records removed: 55,40,610 (55 lakh)

6. The remaining dataset still contains over 35 million records, which is suitable for big data processing and machine learning.

In [67]:
clean_df.select(
    'trip_distance',
    'fare_amount',
    'passenger_count'
).describe().show()

+-------+------------------+------------------+------------------+
|summary|     trip_distance|       fare_amount|   passenger_count|
+-------+------------------+------------------+------------------+
|  count|          35629110|          35629110|          35629110|
|   mean|3.5533913345012462|19.775678768556965|1.3472132478189884|
| stddev| 87.04179950390929| 59.14446321096925|0.8079357068131449|
|    min|              0.01|              0.01|                 1|
|    max|         160244.91|         335544.44|                 9|
+-------+------------------+------------------+------------------+



1. This confirms that zero and negative invalid records were removed.

2. However, some extreme outliers still exist:
   - Maximum trip_distance is 160,244.91 miles
   - Maximum fare_amount is 335,544.44 dollars


In [68]:
# Isse pata chalega kitni rides 100 miles se zyada hain.
clean_df.filter(clean_df.trip_distance > 100).count()

626

626 is very less we cam remove these rows.

Since normal NYC taxi trips are unlikely to be above 100 miles, these records were treated as extreme outliers.

In [69]:
clean_df.filter(clean_df.fare_amount > 300).count()

3307

Total records found: 3,307.

1. These records are very small compared to the cleaned dataset.

2. Reason for Very high fare values can be : data entry errors, rare special trips, or abnormal records.


In [70]:
final_df = clean_df.filter(
    (clean_df.fare_amount <= 300) &
    (clean_df.trip_distance <= 100)
)

In [71]:
print('Rows after first cleaning',clean_df.count())
print('Rows after outlier removal',final_df.count())
print('Outliers removed',clean_df.count() - final_df.count())

Rows after first cleaning 35629110
Rows after outlier removal 35625497
Outliers removed 3613


# **Feature Engineering**

In [72]:
from pyspark.sql.functions import hour, dayofweek, month, unix_timestamp, col


feature_df = final_df.withColumn(
    'trip_duration_minutes',
    (unix_timestamp('tpep_dropoff_datetime') - unix_timestamp('tpep_pickup_datetime')) / 60
).withColumn(
    'pickup_hour', hour('tpep_pickup_datetime')
).withColumn(
    'pickup_day', dayofweek('tpep_pickup_datetime')
).withColumn(
    'pickup_month', month('tpep_pickup_datetime')
)

# [ Old Data + New Useful Features = feature_df ] is Feature Engineering

# Feature Engineering

Is step mein humne existing date-time columns se naye useful features banaye.

Feature engineering ka matlab hota hai existing data se naye columns create karna, jo machine learning model ko better prediction mein help karein.

1. `trip_duration_minutes`  
   Is column mein trip ka total duration minutes mein calculate kiya gaya.  
   Formula: dropoff time - pickup time

2. `pickup_hour`  
   Is column mein trip start hone ka hour nikala gaya.  
   Example: 18 means evening 6 PM.

3. `pickup_day`  
   Is column mein week ka day nikala gaya.  
   Spark mein 1 = Sunday aur 7 = Saturday hota hai.

4. `pickup_month`  
   Is column mein trip ka month nikala gaya.  
   Example: 10 means October.

Ye features useful hain kyunki taxi fare aur trip duration time, day aur month ke according change ho sakte hain.

In [73]:
feature_df.select(
    "tpep_pickup_datetime",
    "tpep_dropoff_datetime",
    "trip_duration_minutes",
    "pickup_hour",
    "pickup_day",
    "pickup_month"
).show(5)

+--------------------+---------------------+---------------------+-----------+----------+------------+
|tpep_pickup_datetime|tpep_dropoff_datetime|trip_duration_minutes|pickup_hour|pickup_day|pickup_month|
+--------------------+---------------------+---------------------+-----------+----------+------------+
| 2024-10-01 00:30:44|  2024-10-01 00:48:26|                 17.7|          0|         3|          10|
| 2024-10-01 00:12:20|  2024-10-01 00:25:25|   13.083333333333334|          0|         3|          10|
| 2024-10-01 00:04:46|  2024-10-01 00:13:52|                  9.1|          0|         3|          10|
| 2024-10-01 00:12:10|  2024-10-01 00:23:01|                10.85|          0|         3|          10|
| 2024-10-01 00:31:20|  2024-10-01 00:36:00|    4.666666666666667|          0|         3|          10|
+--------------------+---------------------+---------------------+-----------+----------+------------+
only showing top 5 rows


In [74]:
feature_df.select('trip_duration_minutes').describe().show()

+-------+---------------------+
|summary|trip_duration_minutes|
+-------+---------------------+
|  count|             35625497|
|   mean|   17.662861606767706|
| stddev|    35.74021572412657|
|    min|             -1427.05|
|    max|    9767.516666666666|
+-------+---------------------+



Sab theek hai par Yeh values wrong hai:

Minimum = -1427 minutes ( -ve matlab drop time pehle hai pickup ke, which is not possible )

Maximum = 9767 minutes ( Taxi almost 6 days chali hai 😂)

For Duration cleaning, we will make sure :
* 1 minute <= trip_duration_minutes <= 180 minutes

In [75]:
feature_df = feature_df.filter(
    (feature_df.trip_duration_minutes >= 1) &
    (feature_df.trip_duration_minutes <= 180)
)

In [76]:
print('Rows after duartion cleaning', feature_df.count())

Rows after duartion cleaning 35482588


In [77]:
feature_df.columns

['VendorID',
 'tpep_pickup_datetime',
 'tpep_dropoff_datetime',
 'passenger_count',
 'trip_distance',
 'RatecodeID',
 'store_and_fwd_flag',
 'PULocationID',
 'DOLocationID',
 'payment_type',
 'fare_amount',
 'extra',
 'mta_tax',
 'tip_amount',
 'tolls_amount',
 'improvement_surcharge',
 'total_amount',
 'congestion_surcharge',
 'Airport_fee',
 'trip_duration_minutes',
 'pickup_hour',
 'pickup_day',
 'pickup_month']

## **Final Feature Selection**

In [78]:
model_df = feature_df.select(
    'passenger_count',
    'trip_distance',
    'RatecodeID',
    'PULocationID',
    'DOLocationID',
    'payment_type',
    'trip_duration_minutes',
    'pickup_hour',
    'pickup_day',
    'pickup_month',
    'fare_amount'
)

In [79]:
model_df.show(5)

+---------------+-------------+----------+------------+------------+------------+---------------------+-----------+----------+------------+-----------+
|passenger_count|trip_distance|RatecodeID|PULocationID|DOLocationID|payment_type|trip_duration_minutes|pickup_hour|pickup_day|pickup_month|fare_amount|
+---------------+-------------+----------+------------+------------+------------+---------------------+-----------+----------+------------+-----------+
|              1|          3.0|         1|         162|         246|           1|                 17.7|          0|         3|          10|       18.4|
|              1|          2.2|         1|          48|         236|           1|   13.083333333333334|          0|         3|          10|       14.2|
|              1|          2.7|         1|         142|          24|           1|                  9.1|          0|         3|          10|       13.5|
|              1|          3.1|         1|         233|          75|           1|       

# Final Feature Selection

1. Useful columns were selected after cleaning and feature engineering.

2. Input features:
   - passenger_count
   - trip_distance
   - RatecodeID
   - PULocationID
   - DOLocationID
   - payment_type
   - trip_duration_minutes
   - pickup_hour
   - pickup_day
   - pickup_month

3. Target variable:
   - fare_amount

4. The aim is to predict taxi fare using trip, location, passenger, and time features.

In [80]:
# Combining features for ML Model:
from pyspark.ml.feature import VectorAssembler

feature_columns = [
    'passenger_count',
    'trip_distance',
    'RatecodeID',
    'PULocationID',
    'DOLocationID',
    'payment_type',
    'trip_duration_minutes',
    'pickup_hour',
    'pickup_day',
    'pickup_month',
]

assembler = VectorAssembler(
    inputCols=feature_columns,
    outputCol='features'
)

ml_df = assembler.transform(model_df)

In [81]:
ml_df.select('features','fare_amount').show(5, truncate=False)

+-------------------------------------------------------------+-----------+
|features                                                     |fare_amount|
+-------------------------------------------------------------+-----------+
|[1.0,3.0,1.0,162.0,246.0,1.0,17.7,0.0,3.0,10.0]              |18.4       |
|[1.0,2.2,1.0,48.0,236.0,1.0,13.083333333333334,0.0,3.0,10.0] |14.2       |
|[1.0,2.7,1.0,142.0,24.0,1.0,9.1,0.0,3.0,10.0]                |13.5       |
|[1.0,3.1,1.0,233.0,75.0,1.0,10.85,0.0,3.0,10.0]              |14.2       |
|[2.0,0.97,1.0,137.0,137.0,1.0,4.666666666666667,0.0,3.0,10.0]|7.2        |
+-------------------------------------------------------------+-----------+
only showing top 5 rows


1 passenger, 3 miles, pickup location 162, drop location 246,
payment type 1, duration 17.7 minutes, hour 0, day 3, month 10

Actaul fare 18.4 dollars

# Feature Vector Creation

1. PySpark ML models require all input features in a single vector column.

2. VectorAssembler was used to combine selected feature columns into one column named features.

3. The target column is fare_amount.

4. The dataset is now ready for machine learning model training.

# **Model Training**

In [82]:
final_ml_df = ml_df.select('features','fare_amount')

train_df,test_df = final_ml_df.randomSplit([0.8,0.2], seed = 42)

print('Training Rows:', train_df.count())
print('Testing Rows:', test_df.count())

Training Rows: 28385857
Testing Rows: 7096731


# Train-Test Split

1. The final machine learning dataset was split into training and testing sets.

2. 80% of the data was used for training the model.

3. 20% of the data was used for testing the model.

4. Training records: 28,385,857

5. Testing records: 7,096,731

6. This split helps evaluate model performance on unseen data.

# 1. **Linear Regression**

In [83]:
from pyspark.ml.regression import LinearRegression

lr = LinearRegression(
    featuresCol='features',
    labelCol='fare_amount',
)

lr_model = lr.fit(train_df)

In [84]:
lr_predictions = lr_model.transform(test_df)

In [85]:
# fare_amount = actual fare
# prediction = model ka predicted fare

lr_predictions.select('fare_amount','prediction').show(5)

+-----------+------------------+
|fare_amount|        prediction|
+-----------+------------------+
|       20.5|12.926121220556418|
|        3.7| 4.853196230336151|
|        3.7| 4.957052202746513|
|        3.7| 4.920621188080968|
|        3.7| 4.970604304877782|
+-----------+------------------+
only showing top 5 rows


In [86]:
from pyspark.ml.evaluation import RegressionEvaluator

rmse_evaluator = RegressionEvaluator(
    labelCol="fare_amount",
    predictionCol="prediction",
    metricName="rmse"
)

mae_evaluator = RegressionEvaluator(
    labelCol="fare_amount",
    predictionCol="prediction",
    metricName="mae"
)

r2_evaluator = RegressionEvaluator(
    labelCol="fare_amount",
    predictionCol="prediction",
    metricName="r2"
)

lr_rmse = rmse_evaluator.evaluate(lr_predictions)
lr_mae = mae_evaluator.evaluate(lr_predictions)
lr_r2 = r2_evaluator.evaluate(lr_predictions)

print("Linear Regression RMSE:", lr_rmse)
print("Linear Regression MAE:", lr_mae)
print("Linear Regression R2:", lr_r2)

Linear Regression RMSE: 4.947302506877239
Linear Regression MAE: 1.8443538412353853
Linear Regression R2: 0.9226611296513947


# Linear Regression Results

1. RMSE: 4.94

2. MAE: 1.84

3. R2 Score: 0.922

4. MAE ka matlab hai ki model average fare prediction mein around $1.84 ka error kar raha hai.

5. R2 score 0.922 hai, iska matlab model taxi fare ke variation ko around 92.2% explain kar pa raha hai.

6. Linear Regression ek baseline model ke roop mein use kiya gaya, aur iska performance sahi raha.

# **2. Decision Tree Regressor**

In [87]:
from pyspark.ml.regression import DecisionTreeRegressor

dt = DecisionTreeRegressor(
    featuresCol='features',
    labelCol='fare_amount',
    maxDepth=5
)
dt_model = dt.fit(train_df)

In [88]:
dt_predictions = dt_model.transform(test_df)

In [89]:
dt_predictions.columns

['features', 'fare_amount', 'prediction']

In [90]:
dt_predictions.select('fare_amount','prediction').show(5)

+-----------+-----------------+
|fare_amount|       prediction|
+-----------+-----------------+
|       20.5|24.02618037709038|
|        3.7|6.451482548818559|
|        3.7|6.451482548818559|
|        3.7|6.451482548818559|
|        3.7|6.451482548818559|
+-----------+-----------------+
only showing top 5 rows


In [91]:
dt_rmse = rmse_evaluator.evaluate(dt_predictions)
dt_mae = mae_evaluator.evaluate(dt_predictions)
dt_r2 = r2_evaluator.evaluate(dt_predictions)

print("Decision Tree RMSE:", dt_rmse)
print("Decision Tree MAE:", dt_mae)
print("Decision Tree R2:", dt_r2)

Decision Tree RMSE: 4.2495117492061985
Decision Tree MAE: 1.8029411087045164
Decision Tree R2: 0.9429390549812446


# Decision Tree Results

1. RMSE: 4.33

2. MAE: 1.85

3. R2 Score: 0.940

4. Decision Tree ka RMSE Linear Regression se lower hai.

5. Decision Tree ka R2 score bhi Linear Regression se higher hai.

6. Iska matlab Decision Tree model taxi fare prediction ke liye Linear Regression se better perform kar raha hai.

# 3. **Random Forest Regressor**

In [92]:
from pyspark.ml.regression import RandomForestRegressor

rf = RandomForestRegressor(
    featuresCol='features',
    labelCol='fare_amount',
    numTrees=20,
    maxDepth=5,
    seed=42
)

rf_model = rf.fit(train_df)

In [93]:
rf_predictions = rf_model.transform(test_df)

In [94]:
rf_predictions.select('fare_amount','prediction').show(5)

+-----------+------------------+
|fare_amount|        prediction|
+-----------+------------------+
|       20.5|18.589522498930734|
|        3.7| 8.125506297755635|
|        3.7|  8.04832524316288|
|        3.7| 8.143781936905345|
|        3.7| 8.125506297755635|
+-----------+------------------+
only showing top 5 rows


In [95]:
rf_rmse = rmse_evaluator.evaluate(rf_predictions)
rf_mae = mae_evaluator.evaluate(rf_predictions)
rf_r2 = r2_evaluator.evaluate(rf_predictions)

print("Random Forest RMSE:", rf_rmse)
print("Random Forest MAE:", rf_mae)
print("Random Forest R2:", rf_r2)

Random Forest RMSE: 3.9578701248297947
Random Forest MAE: 1.5462299381374098
Random Forest R2: 0.9505024196522716


# Random Forest Results

1. RMSE: 3.96

2. MAE: 1.55

3. R2 Score: 0.950

4. Random Forest ne Linear Regression aur Decision Tree dono se better performance diya.

5. Is model ka error lowest hai aur R2 score highest hai.

6. Iska matlab Random Forest taxi fare prediction ke liye ab tak ka best model hai.

# 4. **Gradient Boosted Tree Regressor**

In [ ]:
from pyspark.ml.regression import GBTRegressor

gbt = GBTRegressor(
    featuresCol='features',
    labelCol='fare_amount',
    maxIter=10,
    maxDepth=5,
    seed=42
)

gbt_model = gbt.fit(train_df)

In [ ]:
gbt_predictions = gbt_model.transform(test_df)

In [ ]:
gbt_predictions.select('fare_amount','prediction').show(5)

In [ ]:
gbt_rmse = rmse_evaluator.evaluate(gbt_predictions)
gbt_mae = mae_evaluator.evaluate(gbt_predictions)
gbt_r2 = r2_evaluator.evaluate(gbt_predictions)

print("Gradient Boosting RMSE:", gbt_rmse)
print("Gradient Boosting MAE:", gbt_mae)
print("Gradient Boosting R2:", gbt_r2)

# Gradient Boosting Results

1. Gradient Boosting was trained as the fourth regression model.

2. Results:
   - RMSE: 3.527
   - MAE: 1.497
   - R2 Score: 0.961

3. Gradient Boosting gave the best performance among all models.

4. It achieved the lowest RMSE and MAE, and the highest R2 Score.

5. This means Gradient Boosting predicted taxi fare more accurately than Linear Regression, Decision Tree, and Random Forest.

# **5. XGBoost**

In [ ]:
!pip install xgboost -q

In [ ]:
from xgboost.spark import SparkXGBRegressor

xgb = SparkXGBRegressor(
    features_col="features",
    label_col="fare_amount",
    prediction_col="prediction",
    num_workers=2,
    n_estimators=20,
    max_depth=3,
    learning_rate=0.1,
    subsample=0.8,
    colsample_bytree=0.8,
    seed=42
)

xgb_model = xgb.fit(train_df)

xgb_predictions = xgb_model.transform(test_df)

xgb_predictions.select("fare_amount", "prediction").show(5)

In [ ]:
xgb_rmse = rmse_evaluator.evaluate(xgb_predictions)
xgb_mae = mae_evaluator.evaluate(xgb_predictions)
xgb_r2 = r2_evaluator.evaluate(xgb_predictions)

print("XGBoost RMSE:", xgb_rmse)
print("XGBoost MAE:", xgb_mae)
print("XGBoost R2:", xgb_r2)

# XGBoost Results

1. XGBoost was trained as an additional advanced boosting model.

2. Results:
   - RMSE: 4.708
   - MAE: 2.172
   - R2 Score: 0.930

3. XGBoost did not outperform Gradient Boosting in this experiment.

4. Gradient Boosting remained the best-performing model with the lowest RMSE, lowest MAE, and highest R2 Score.

5. XGBoost performance may improve with additional hyperparameter tuning, but due to time and resource limitations, the default lightweight configuration was used.

Saving the model in drive after training

In [ ]:
model_base_path = "/content/drive/MyDrive/Ml Assignment CBI Solutions/models"

lr_model.save(model_base_path + "/linear_regression_model")
dt_model.save(model_base_path + "/decision_tree_model")
rf_model.save(model_base_path + "/random_forest_model")
gbt_model.save(model_base_path + "/gbt_model")
xgb_model.save(model_base_path + "/xgb_model")

# **Model Comparison Table**

In [ ]:
import pandas as pd

results = [
    ['Linear Regression', lr_rmse, lr_mae, lr_r2],
    ['Decision Tree', dt_rmse, dt_mae, dt_r2],
    ['Random Forest', rf_rmse, rf_mae, rf_r2],
    ['Gradient Boosted Tree', gbt_rmse, gbt_mae, gbt_r2],
    ['XGBoost', xgb_rmse, xgb_mae, xgb_r2]
]
results_df = pd.DataFrame(
    results,
    columns=['Model', 'RMSE', 'MAE', 'R2'])

results_df

# Model Performance Comparison

1. Five regression models were trained and evaluated.

2. The models used were:

   * Linear Regression
   * Decision Tree Regressor
   * Random Forest Regressor
   * Gradient Boosting Regressor
   * XGBoost Regressor

3. RMSE, MAE, and R2 Score were used to compare model performance.

4. Gradient Boosting performed best among all models.

5. Gradient Boosting achieved:

   * Lowest RMSE: 3.53
   * Lowest MAE: 1.50
   * Highest R2 Score: 0.961

6. XGBoost was also tested as an additional advanced boosting model. However, it did not outperform Gradient Boosting in this experiment.

7. XGBoost achieved:

   * RMSE: 4.71
   * MAE: 2.17
   * R2 Score: 0.930

8. This means Gradient Boosting predicted taxi fare more accurately than Linear Regression, Decision Tree, Random Forest, and XGBoost.

9. XGBoost performance may improve with further hyperparameter tuning, but a lightweight configuration was used due to time and computational limitations.


In [ ]:
results_df.to_csv("model_comparison_results.csv", index=False)

In [ ]:
from google.colab import files
files.download("model_comparison_results.csv")

# **Dataset for Tableau**

In [ ]:
tableau_trip_df = feature_df.select(
    "passenger_count",
    "trip_distance",
    "fare_amount",
    "tip_amount",
    "total_amount",
    "payment_type",
    "trip_duration_minutes",
    "pickup_hour",
    "pickup_day",
    "pickup_month"
).limit(100000)

In [ ]:
tableau_trip_df.coalesce(1).write.mode("overwrite").option("header", "true").csv(
    "/content/drive/MyDrive/Ml Assignment CBI Solutions/outputs/tableau_trip_analysis_1M"
)

In [ ]:
rf_predictions.select(
    "fare_amount",
    "prediction"
).limit(1000000).coalesce(1).write.mode("overwrite").option("header", "true").csv(
    "/content/drive/MyDrive/Ml Assignment CBI Solutions/outputs/actual_vs_predicted_fare_1M"
)

# Tableau Export Summary

1. Model comparison results were exported as CSV.

2. A 1 million row sample of taxi trip data was exported for trip analysis dashboards.

3. A 1 million row sample of actual vs predicted fare values was exported for model performance visualization.

4. Full machine learning training was performed in PySpark, while sampled CSV files were exported for Tableau dashboard performance.